Adaptive GC Threading

RSalman opened the pull request:
#### **_For background and results see https://github.com/eclipse/omr/issues/5829_** 

> Using overhead data (busy/stall times for managing and synchronizing threads), Adaptive Threading aims to identify sub-optimal/detrimental parallelism and continuously adjusts the GC thread count to reach optimal parallelism.

_Adaptive Threading changes are implemented at the various phases of GC as follows, during:_

1. **Pre-collection _(during to task/thread dispatch)_:** to adjust thread count based on the previous cycle’s recommendation
2. **Collection:** to gather data for parallelization overhead for the running cycle
3. **Post-collection _(after worker threads are suspended)_:** to project/calculate optimal thread count and give recommendation for the next cycle based on the current GC that’s completed

 Adaptive threading will be enabled by default. The user may choose to enable/disable adaptive threading through the `-XX:[+-]AdaptiveGCThreading` options. However Adaptive Threading is ignored, even if it is enabled, when GC thread count is forced (e.g user specifics Xgcthreads). The user can also specify upper thread limit for adaptive threading using  `-Xgcmaxthreads` option.

_The specifics of the implementation are as follows:_

#### Introduced concept of recommended threads:
 -  Thread count associated with a parallel task instance, indicates the projected optimal number of threads to complete the task. 
 - Drives Adaptive Threading - changes with each dispatch of the task, based on observations made when previously completing the same task
 - Currently, adaptive threading is only used with  _Scavenger_,  hence this is only applies to tasks of type `MM_ParallelScavengeTask` so a recommended thread count is provided with a new `ParallelScavengeTask` instance (i.e when scavenge is run)
 -  `getRecommendedWorkingThreads` introduced to `MM_Task` base class, base implementation returns `UDATA_MAX` (signifies no adaptive threading), overridden  by `ParallelScavengeTask` to return adaptive thread count.

#### Introduced the following ScavengerStats (New set of stats gathered during collection):
- ` _workerScavengeStartTime` and `_workerScavengeEndTime`
  - Thread local task (scavenge) start/end timestamps taken when worker thread starts/competes a task (scavenge). This is used to determine the time it takes a worker to start collection task from the time a cycle starts and how long a worker waits for others when it is completed its task.
- `_adjustedSyncStallTime`
  - Similar to the existing `_SyncStallTime` stat **except** it accounts for Critical Section duration. That is, it subtracts   critical section duration from stall time of synced thread. This is because the stall time being added from critical section is independent of number of the threads being synchronized. This independent stall time can not be adjusted for, hence it must be ignored. This is relevant for `syncAndRealeaseSingle/Master` APIs as they are the only APIs that record a stall time with critical sections, without these, `_SyncStallTime` == `_adjustedSyncStallTime`
  - existing `addToSyncStallTime` method extended  to update `_adjustedSyncStallTime` in addition to `_SyncStallTime`.  `addToSyncStallTime` now takes  `criticalSectionDuration` (defaults to 0) as an input, a value for it passed when a critical section is executed prior to updating stall stats .
 - `_notifyStallTime `
   - Used to record the stall times resulting from notifying other waiting threads. Note, this stat is not inclusive, it only records notify times relevant to adaptive threading.
   - We are more concerned about recording 'notify_all' rather than 'notify_one' as 'notify_all' is dependent on the number of threads being notified.

#### Collector changes, Scavenger updates:
- Introduced `calculateRecommendedWorkingThreads` routine
  - Called at the end of each successful scavenge to project optimal threads for next scavenge
  - Implements Adaptive Model _(see background for more info)_
  - Calculates averages of stall stats and uses them as inputs to adaptive model, projected optimal thread count output is stored to new member `_recommendedThreads`.
- Timestamps taken for worker thread scavenge start/end time
 - `notifyStallTime` recorded  for `notify_all` on `_scanCacheMonitor` _(some instances are ignored as they are not relevant to adaptive threading such as backout cases)_
- `MergeThreadGCStats` updated to merge new scavenger stats (discussed above)
  - trace point added here to breakdown thread local stats for adaptive threading

#### Dispatcher's `recomputeActiveThreadCountForTask` updated to account for Adaptive Threading
- Dispatcher now queries the task for the recommended thread count when determining the number of threads to release from thread pool to complete the task
  - Ensures thread count is bounded properly to respect max thread count, either user provided adaptiveThreadCount or default max.

#### Synchronization API updates:
- `_syncCriticalSectionStartTime` & `_syncCriticalSectionDuration` introduced to record Critical Section times for adjusted stall time
  -  `_syncCriticalSectionStartTime` recorded when all threads are synced and the released thread is about to exit sync api to execute critical section
  - `_syncCriticalSectionDuration` is recored when thread executing critical section is about to release the synced thread (indicating critical section is complete). As synced threads are released, they update their stall time with the newly set critical section duration.
- `notifyStallTime` recorded  for `notify_all` on synced threads  

Signed-off-by: Salman Rana <salman.rana@ibm.com>

RSalman commented:
cc @amicic @dmitripivkine

RSalman commented:
Out of date files, uncommitted files causing PR build check fails. I'll fix this

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -457,6 +457,26 @@ MM_ParallelDispatcher::recomputeActiveThreadCountForTask(MM_EnvironmentBase *env
+	 */
+	if (_extensions->adaptiveThreadigEnabled() && task->getRecommendedWorkingThreads() != UDATA_MAX) {
+		/* Bound the recommended thread count. Determine the  upper bound for the thread count,
+		 * either the user specified adaptiveThreadCount (-XXgc:adaptiveThreads=) or else default max
```
I think we should make the option -Xgc: (single X, documented, supported)

RSalman replied:
Changed - see https://github.com/eclipse/openj9/pull/11999

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -457,6 +457,26 @@ MM_ParallelDispatcher::recomputeActiveThreadCountForTask(MM_EnvironmentBase *env
+		 */
+		uintptr_t upperThreadBound = (_extensions->adaptiveThreadCount == 0) ? _threadCountMaximum : _extensions->adaptiveThreadCount;
+		taskActiveThreadCount = OMR_MIN(upperThreadBound, task->getRecommendedWorkingThreads());
+		taskActiveThreadCount = OMR_MAX(1, taskActiveThreadCount);
```
perhaps it's cleaner if we make getRecommendedWorkingThreads returns >= 1 (if that's the reason we do this  OMR_MAX(1,...)

RSalman replied:
recommended thread will always return  >= 1,
https://github.com/RSalman/omr/blob/893eb8f9413eb74eecdd8135ea4b62e115ab019f/gc/base/standard/Scavenger.cpp#L567-L569

 so this isn't needed

amicic commented on gc/base/standard/ParallelScavengeTask.cpp:
```diff
@@ -88,7 +88,15 @@ MM_ParallelScavengeTask::synchronizeGCThreadsAndReleaseMain(MM_EnvironmentBase *
+	}
+
+	/* _syncCriticalSectionDuration must be set now, this thread's stall time is at least the duration of critical section */
+	Assert_MM_true((endTime - startTime) >= _syncCriticalSectionDuration);
```
Since _syncCriticalSectionStartTime/Duration are introuced by base class (and indeed in theory some other Tasks could benefit from it), It would be nice if we could move all this new logic into the base class method. Then the existing code that sets start/endTime should move too, and it's possible if we hold them in Env rather than on stack. Many other Parallel Tasks could be adjusted to that paradigm, too.

The line below that updates the stats could either stay here, or even be wrapped into a virtual method and go inside base too (in which case, whole this method would not need to be virtualized and specialized for each Parallel Task).

This is ok for now, but something to think about.

amicic commented on gc/stats/ScavengerStats.hpp:
```diff
@@ -98,6 +98,17 @@ class MM_ScavengerStats
 
+	/* Stats Used Specifically for Adaptive Threading */
+	uint64_t _workerScavengeStartTime; /**< Timestamp taken when worker starts the scavenge task */
+	uint64_t _workerScavengeEndTime; /**< Timestamp taken when worker completes the scavenge task */
```
I wish there are simple called _start/_endTime (this is Thread local and Scavenge specific class already, and workerScavenge is redundant). 

But I understand that those already exist, although they are somewhat poorly used - not as thread local ones, but to record total cycle duration, done only my main thread. Ideally, the existing ones should be moved somewhere else (for example in Collector class), but that would required some significant changes (how Verbose GC obtains them).

For now, this is good, but something to think about. At least, please, add a comment to existing ones why they are used for.

amicic commented on gc/base/standard/Scavenger.cpp:
```diff
@@ -479,14 +485,100 @@ MM_Scavenger::calculateMaxCacheCount(uintptr_t activeMemorySize)
+	/* Calculate the average time it takes the worker threads to start collection and avgerage time workers are idle waiting for task cleanup
+	 * Calculated as (Sum_WorkerStartTime(t1 + t2 + ... + tn) - (n * collection_start)) / n  */
+	uint64_t avgTimeToStartCollection =  omrtime_hires_delta((_extensions->scavengerStats._startTime * _dispatcher->activeThreadCount()), _extensions->scavengerStats._workerScavengeStartTime, OMRPORT_TIME_DELTA_IN_MICROSECONDS) / totalThreads;
+	uint64_t avgTimeIdleAfterCollection =  omrtime_hires_delta(_extensions->scavengerStats._workerScavengeEndTime, (_extensions->scavengerStats._endTime * _dispatcher->activeThreadCount()), OMRPORT_TIME_DELTA_IN_MICROSECONDS) / totalThreads;
```
in this block you have activeThreadCount() while you can use local totalThreads

amicic commented on gc/base/standard/Scavenger.cpp:
```diff
@@ -479,14 +485,100 @@ MM_Scavenger::calculateMaxCacheCount(uintptr_t activeMemorySize)
+	double powerBase = (1.0 / modifier) * stallModifider;
+
+	double idealThreads = _dispatcher->activeThreadCount() * pow(powerBase, (1 / powerModifier));
+	float adjustedAverage = MM_Math::weightedAverage((float)_dispatcher->activeThreadCount(), (float)idealThreads, _extensions->adaptiveThreadingWeightActiveThreads);
```
You can introduce new API for double to avoid casting. Although the whole preceding calculation and all params could've been floats, too.

RSalman replied:
Fixed - changed all preceding calculations and params to float. With this changed, the extensions variables are changed to float as well.

RSalman replied:
https://github.com/RSalman/omr/blob/893eb8f9413eb74eecdd8135ea4b62e115ab019f/gc/base/GCExtensionsBase.hpp#L1619-L1621

RSalman replied:
https://github.com/RSalman/omr/blob/893eb8f9413eb74eecdd8135ea4b62e115ab019f/gc/base/GCExtensionsBase.hpp#L1619-L1621

amicic commented on gc/base/GCExtensionsBase.hpp:
```diff
@@ -1093,6 +1101,22 @@ class MM_GCExtensionsBase : public MM_BaseVirtual {
+		return (!concurrentScavenger && adaptiveThreadingEnabled);
+#else /* defined(OMR_GC_CONCURRENT_SCAVENGER) */
+		return adaptiveThreadingEnabled;
+#endif /* defined(OMR_GC_CONCURRENT_SCAVENGER) */
```
there is a helper isConcurrentScavengerEnabled() that already deals with ifdefs and can make this simpler

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -457,6 +457,25 @@ MM_ParallelDispatcher::recomputeActiveThreadCountForTask(MM_EnvironmentBase *env
+	 *  2) Adaptive threading flag is not set (-XX:-AdaptiveGCThreading)
+	 *  3) The task wasn't recommended a thread count (currently only recommended for STW Scavenge Tasks)
+	 */
+	if (_extensions->adaptiveThreadigEnabled() && task->getRecommendedWorkingThreads() != UDATA_MAX) {
```
extra bracket

amicic commented on gc/base/GCExtensionsBase.hpp:
```diff
@@ -1595,6 +1619,11 @@ class MM_GCExtensionsBase : public MM_BaseVirtual {
+		, adaptiveThreadCount(0)
+		, adaptiveThreadingSensitivityFactor((float)1.0)
+		, adaptiveThreadingWeightActiveThreads((float)0.50)
+		, adaptiveThreadBooster((float)0.85)
```
I believe more correct is to do 0.85f

RSalman replied:
changed - I prefer this as well. It's just that I saw the casting style done in the file.

RSalman commented:
`x86-64 Windows` is giving troubles with types _(warning treated as error)_ even after changing all calculations/vars types to float. This is why I’ve added these 2 explicit casts :

```
float idealThreads = (float)(totalThreads * pow(powerBase, powerExponent));

float adjustedAverage = MM_Math::weightedAverage((float)totalThreads, idealThreads, _extensions->adaptiveThreadingWeightActiveThreads);
```

amicic commented:
there is powf variant to avoid first case of casting.
for second one, yeah, you'll need it

RSalman commented:
A few changes made since last review:
- Enable Adaptive Threading By default
- Removed adaptiveGCThreadCount from extensions (the existing gc thread count variable will store this)
-  Parallel Dispatcher Bounding of thread count no longer looks to adaptiveGCThreadCount
- Added comment to `_startTime` & `_endTime` of Scavenger Stats (as suggested)
-  Extension's `adaptiveThreadigEnabled()` no longer considers Concurrent Scavenger, this API should be collector independent... As a result,`Scavenger::calculateRecommendedWorkingThreads` now explicitly checks if CS is enabled to skip the routine. When Parallel Dispatcher quires `adaptiveThreadigEnabled()`, it doesn't need to check for CS as the recommendeWorkingThreads will return MAX and fail anyways.

amicic commented on gc/base/standard/Scavenger.cpp:
```diff
@@ -479,14 +485,100 @@ MM_Scavenger::calculateMaxCacheCount(uintptr_t activeMemorySize)
+void
+MM_Scavenger::calculateRecommendedWorkingThreads(MM_EnvironmentStandard *env)
+{
+	if (!_extensions->adaptiveThreadigEnabled() || _extensions->isConcurrentScavengerEnabled()) {
```
It might be overkill to check if adaptive is enabled both here and in dispatcher!?

It would be interesting to calculate recommended value  (and report through trace point)  even is adaptive is disabled.

But then, in a case  of this calculation misbehaving (for example div by 0), it's good to have ability to disable it. So, let's play safe and leave the check here (and in Dispatcher too).

RSalman replied:
I agree, it could be interesting to calculate the recommended value regardless. But we should keep the ability to skip the routine in case of "anything". 

By leaving the check in Scavenger,  the check in Dispatcher is defiantly an overkill:

`if (_extensions->adaptiveThreadigEnabled() && (task->getRecommendedWorkingThreads() != UDATA_MAX)) `

`getRecommendedWorkingThreads` will always cause condition to fail unless adaptive threading is enabled. Perhaps this can be simplified

RSalman commented:
@fjeremic could you please have a look, this is ready to be merged/further review

fjeremic commented:
Jenkins build all

RSalman commented:
CI fails seem like an infra issue... 

**continuous-integration/eclipse-omr/pr/win_x86-64** 
```
15:57:58  100% tests passed, 0 tests failed out of 23
15:57:58  
15:57:58  Total Test time (real) = 108.67 sec
```
_Failing because it can't delete the workspace:_
```
ERROR: Cannot delete workspace: Unable to delete 'c:\jenkins\workspace\PullRequest-win_x86-64\Build'. Tried 3 times (of a maximum of 3) waiting 0.1 sec between attempts.
Finished: FAILURE
```



**continuous-integration/eclipse-omr/pr/osx_x86-64**
```
15:48:08  100% tests passed, 0 tests failed out of 46
15:48:08  
15:48:08  Total Test time (real) = 254.77 sec
```
_Failing because of some IO issue_
```
Setting status of 403d231b3b492a216eb64620ff6f412d6d0147d5 to FAILURE with url https://ci.eclipse.org/omr/job/PullRequest-osx_x86-64/2273/ and message: 'Build finished. '
Using context: continuous-integration/eclipse-omr/pr/osx_x86-64
Also:   hudson.model.Computer$TerminationRequest: Termination requested at Thu Mar 18 15:54:28 EDT 2021 by Thread[Ping thread for channel hudson.remoting.Channel@79a492cc:osx1013-unb-01,5,main] [id=343]
...
java.io.IOException: Unable to create live FilePath for osx1013-unb-01
...
Finished: FAILURE
```

RSalman commented:
Perhaps the CI checks can be restarted?

fjeremic commented:
We have coverage for the two failures from Azure.
